### PR TITLE
Allow message prefix - #544

### PR DESCRIPTION
**Pull request by rluvaton**
Allowing message prefix

fixing:
- #544

This PR contains:

- [x] Fix
- [x] Tests
- [x] Updated docs

### CODE REVIEW

**Comment by rluvaton** [comment]
Could someone please approve running the workflows so I can see the performance change :)

**Comment by rluvaton** [comment]
Could someone please stop this job:
https://github.com/pinojs/pino/actions/runs/4049806965/jobs/6977390174

**Comment by rluvaton** [comment]
@Fdawgs Could you please me out and allow running the workflows 😬

**Comment by Fdawgs** [comment]
Have reran it @rluvaton

**Comment by mcollina** [comment]
The benchmarks looks ok to me. Could you add some tests for the fix you are making?

**Comment by rluvaton** [comment]
yeah sure, to make sure this fix is ok for you, I'll try to add it + docs later this day

File: lib/tools.js [hunk 1]
```diff
@@ -52,9 +53,16 @@ function genLog (level, hook) {
         msg = n.shift()
         formatParams = n
       }
+      if (typeof this[msgPrefixSym] === 'string' && msg !== undefined && msg !== null) {
+        msg = this[msgPrefixSym] + msg
```
Comment (jsumners) [inline lib/tools.js:57]: I think this is one of the cases where a coercive check is preferred:

```suggestion
      if (typeof this[msgPrefixSym] === 'string' && msg) {
        msg = this[msgPrefixSym] + msg
```

File: lib/tools.js [hunk 2]
```diff
@@ -52,9 +53,16 @@ function genLog (level, hook) {
...
+      }
       this[writeSym](o, format(msg, formatParams, this[formatOptsSym]), level)
     } else {
-      this[writeSym](null, format(o === undefined ? n.shift() : o, n, this[formatOptsSym]), level)
+      let msg = o === undefined ? n.shift() : o
+      if (typeof this[msgPrefixSym] === 'string' && msg !== undefined && msg !== null) {
+        msg = this[msgPrefixSym] + msg
```
Comment (jsumners) [inline lib/tools.js:63]:
```suggestion
      if (typeof this[msgPrefixSym] === 'string' && msg) {
        msg = this[msgPrefixSym] + msg
```

File: lib/tools.js (hunk 1 above)
Comment (rluvaton) [reply lib/tools.js]: thanks, done

File: lib/tools.js (hunk 2 above)
Comment (rluvaton) [reply lib/tools.js]: thanks, done

**Comment by rluvaton** [comment]
This PR is now ready for review @jsumners

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by rluvaton** [comment]
let's merge it?

**Comment by mcollina** [comment]
let's wait for @jsumners

File: docs/api.md [hunk 3]
```diff
@@ -842,6 +842,23 @@ const child = logger.child({foo: 'bar'}, {level: 'debug'})
 child.debug('debug!') // will log as the `level` property set the level to debug
 ```
 
+##### `options.msgPrefix` (String)
+
+The `msgPrefix` property allows you to specify a prefix for every message of the child logger.
+By default, the parent prefix is inherited.
+If the parent already has a prefix, the prefix of the parent and then the child will be displayed.
+
```
Comment (jsumners) [inline docs/api.md:850]: I think this is also a root logger option, correct? We should have documentation that explains the option for the root logger. I think we are also missing a test that covers the root logger.

Comment (rluvaton) [reply docs/api.md:850]: no, it's child logger option

Comment (rluvaton) [reply docs/api.md]: @jsumners Added support for the root logger now

File: docs/api.md [hunk 4]
```diff
@@ -384,6 +384,24 @@ See `errorKey` option to change `err` namespace.
 
 * See [pino.stdSerializers](#pino-stdserializers)
 
+#### `msgPrefix` (String)
+
+Default: `undefined`
+
+The `msgPrefix` property allows you to specify a prefix for every message of the logger and it's children.
```
Comment (jsumners) [inline docs/api.md:391]:
```suggestion
The `msgPrefix` property allows you to specify a prefix for every message of the logger and its children.
```

**Review by jsumners — APPROVED** [review]
Looks good to me.

**Comment by rluvaton** [comment]
I think the failed test is flaky, had it before... in the CI but couldn't reproduce on ubuntu, would love a rerun

**Comment by jsumners** [comment]
I don't see any errors with the CI run. But the benchmarks may not be acceptable. @mcollina can you run them on your dedicated system?

**Comment by rluvaton** [comment]
They changed a lot between:
504d1f1bdc60ba492245d6553a160cd84ed4e9db

to

f8067843fdc2b11bfcd3ad6576ae8c88d94425a1

and I only changed this:
[`a94a9da` (#1635)](https://github.com/pinojs/pino/pull/1635/commits/a94a9da3166672d2857650c7d2dc0f1d4153f12f#diff-7f88797c7db38a6af7a3283e3af30f2cdd3c2489bd9c85a7a25b32ec2485f724R56-R62)

**Comment by jsumners** [comment]
> They changed a lot between: [504d1f1](https://github.com/pinojs/pino/commit/504d1f1bdc60ba492245d6553a160cd84ed4e9db)
> 
> to
> 
> [f806784](https://github.com/pinojs/pino/commit/f8067843fdc2b11bfcd3ad6576ae8c88d94425a1)
> 
> and I only changed this: [`a94a9da` (#1635)](https://github.com/pinojs/pino/pull/1635/commits/a94a9da3166672d2857650c7d2dc0f1d4153f12f#diff-7f88797c7db38a6af7a3283e3af30f2cdd3c2489bd9c85a7a25b32ec2485f724R56-R62)

Then the coercive check may not be acceptable. Revert the change and add a note like:

```js
// We do not use a coercive check for `msg` as it is
// measurably slower than the explicit checks.
```

**Comment by rluvaton** [comment]
Reverted but it doesn't make any sense, this check has never met as we don't run benchmarks with log-prefix

**Comment by rluvaton** [comment]
it's ready for merge :)

**Comment by rluvaton** [comment]
@jsumners can this be merged?

**Pull request merged**